spec/utils_spec.lua

```diff
@@ -146,3 +146,164 @@ describe("utils.validateUndername", function()
+	end)
+
+	it("should handle non-existent records", function()
+		assert.has_error(function()
+			utils.assertHasRecordPermission("ZQhZNfK0JvzJCq-6jqJ4OWt6jwwDcxpgD5lPxOx3VEU", "nonexistent")
+		end, "Sender does not have permission for this record.")
+	end)
+
+	it("should handle records with nil owner", function()
+		Records["owned"].owner = nil
+
+		-- Only ANT-level permissions should work
+		assert.has_no.error(function()
+			utils.assertHasRecordPermission("BaMK_9bFWZMsRmk1L5h0UgO0p-xh_oUpqdCCPQhJlkI", "owned")
+		end)
+
+		assert.has_no.error(function()
+			utils.assertHasRecordPermission("nX5LnVMl1sHs3Og7rU6nQDKSNVWKdMG_Y8dgHdXEqCw", "owned")
+		end)
+
+		-- Record owner should fail since owner is nil
+		assert.has_error(function()
+			utils.assertHasRecordPermission("Th2GyXvBSav3fV6I_4RgjV-xXJZnN2LNYnzgIJQxgKg", "owned")
+		end, "Sender does not have permission for this record.")
+	end)
+
+	it("should check ANT permissions before record permissions", function()
+		-- Even if someone is a record owner of one record,
+		-- if they're a controller, they can modify all records
+		_G.Controllers = { "Th2GyXvBSav3fV6I_4RgjV-xXJZnN2LNYnzgIJQxgKg" }
+
+		assert.has_no.error(function()
+			utils.assertHasRecordPermission("Th2GyXvBSav3fV6I_4RgjV-xXJZnN2LNYnzgIJQxgKg", "unowned")
+		end)
+	end)
+
+	it("should allow owner via Balances check", function()
+		-- Test the Balances check for ownership
+		Balances["rpL3v0QfGaNqb-_6TW9KfP7kQPPM0j_R9_5l2FdakBs"] = 1
+
+		assert.has_no.error(function()
+			utils.assertHasRecordPermission("rpL3v0QfGaNqb-_6TW9KfP7kQPPM0j_R9_5l2FdakBs", "owned")
+		end)
+	end)
+
+	it("should deny access when Balances is 0", function()
+		-- Test that balance of 0 doesn't grant access
+		Balances["3lF8R9cKOPx2E6xFdh1JQXQ7hVRzb_tCi5k5GkLP2cs"] = 0
+
+		assert.has_error(function()
+			utils.assertHasRecordPermission("3lF8R9cKOPx2E6xFdh1JQXQ7hVRzb_tCi5k5GkLP2cs", "owned")
+		end, "Sender does not have permission for this record.")
+	end)
+
+	it("should deny access when Balances is non-1 value", function()
+		-- Test that other balance values don't grant access
+		Balances["qPHOj3S4L7AxZW6NaKu_p1QFMzJsYK9zYLvr4KkNi2w"] = 2
+
+		assert.has_error(function()
+			utils.assertHasRecordPermission("qPHOj3S4L7AxZW6NaKu_p1QFMzJsYK9zYLvr4KkNi2w", "owned")
+		end, "Sender does not have permission for this record.")
+
+		-- Test with decimal
+		Balances["HvaMT3gGlmR7F3bFCBKqE5B0yC8jQMKwQwGsULQVYpw"] = 0.5
+
+		assert.has_error(function()
+			utils.assertHasRecordPermission("HvaMT3gGlmR7F3bFCBKqE5B0yC8jQMKwQwGsULQVYpw", "owned")
+		end, "Sender does not have permission for this record.")
+	end)
+end)
+
+describe("undernameForName", function()
+	it("should get the undername name for a name with an undername", function()
+		local undername = "undername"
+		local basename = "basename"
+		local name = undername .. "_" .. basename
+
+		local undernameFromName = utils.undernameForName(name)
+		assert.are.same(undernameFromName, undername)
```

Comment:
🤔 - what is the intent of this function? base names wouldn't be known by ANTs

Reply:
the primary name workflows get called with the arns "fqdn" so it comes with the base name attached. We dont particularly care about the base name, but we need to extract the undername